File: app/src/main/java/com/billpayments/features/payments/data/PaymentsStore.kt
package com.billpayments.features.payments.data

import com.billpayments.core.event.UiEvent
import com.billpayments.core.event.UiEventBus
import com.billpayments.core.time.Clock
import com.billpayments.features.auth.data.AuthStore
import com.billpayments.features.auth.data.userIdOrNull
import com.billpayments.features.bills.data.DueBillsStore
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.drop
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch

@Singleton
class PaymentsStore @Inject constructor(
    authStore: AuthStore,
    private val paymentRepository: PaymentRepository,
    private val historyRepository: PaymentHistoryRepository,
    private val dueBillsStore: DueBillsStore,
    private val events: UiEventBus,
    private val clock: Clock,
    private val scope: CoroutineScope,
) {
    private var epoch = 0
    private var userId: String? = authStore.state.value.userIdOrNull
    private val _state = MutableStateFlow(userId?.let(historyRepository::restore) ?: Payments.EMPTY)
    val state: StateFlow<Payments> = _state.asStateFlow()

    init {
        scope.launch {
            authStore.state.map { it.userIdOrNull }.distinctUntilChanged().drop(1).collect { id ->
                epoch++
                userId = id
                _state.value = if (id == null) Payments.EMPTY else historyRepository.restore(id)
            }
        }
        scope.launch {
            _state.drop(1).collect { next ->
                val id = userId ?: return@collect
                try {
                    historyRepository.persist(id, next)
                } catch (_: Exception) {
                    events.emit(UiEvent.StorageFailed)
                }
            }
        }
    }

    fun pay(billerId: String, billerName: String, categoryId: String, account: String, amountPaise: Long) {
        if (_state.value.hasProcessing(billerId, account)) return
        val payment = Payment(
            id = "pay-${_state.value.nextId}",
            billerId = billerId,
            billerName = billerName,
            categoryId = categoryId,
            account = account,
            amountPaise = amountPaise,
            paidAtUtc = clock.now(),
            status = PaymentStatus.Processing,
        )
        _state.value = _state.value.adding(payment)
        events.emit(UiEvent.PaymentStarted(payment.id))
        val startedEpoch = epoch
        scope.launch {
            try {
                paymentRepository.pay(payment)
                if (startedEpoch != epoch) return@launch
                _state.value = _state.value.updatingStatus(payment.id, PaymentStatus.Success)
                dueBillsStore.invalidate()
            } catch (_: Exception) {
                if (startedEpoch != epoch) return@launch
                _state.value = _state.value.updatingStatus(payment.id, PaymentStatus.Failed)
                events.emit(UiEvent.PaymentFailed(payment.id))
            }
        }
    }
}

File: BillPayments/Features/Payments/Data/PaymentsStore.swift
import Observation

@Observable
final class PaymentsStore {
    private(set) var state: Payments
    @ObservationIgnored private(set) var pendingWork: Task<Void, Never>?
    @ObservationIgnored private(set) var observationWork: Task<Void, Never>?

    @ObservationIgnored private let paymentRepository: PaymentRepository
    @ObservationIgnored private let historyRepository: PaymentHistoryRepository
    @ObservationIgnored private let invalidateDueBills: () -> Void
    @ObservationIgnored private let events: UiEventBus
    @ObservationIgnored private let clock: Clock
    @ObservationIgnored private var epoch = 0
    @ObservationIgnored private(set) var userId: String?

    init(
        authStore: AuthStore,
        paymentRepository: PaymentRepository,
        historyRepository: PaymentHistoryRepository,
        invalidateDueBills: @escaping () -> Void,
        events: UiEventBus,
        clock: Clock
    ) {
        self.paymentRepository = paymentRepository
        self.historyRepository = historyRepository
        self.invalidateDueBills = invalidateDueBills
        self.events = events
        self.clock = clock
        if let id = authStore.state.userIdOrNull {
            self.userId = id
            self.state = historyRepository.restore(userId: id)
        } else {
            self.state = .empty
        }
        observationWork = Task { [weak self, authStore] in
            for await userId in Observations({ authStore.state.userIdOrNull }) {
                guard let self else { return }
                self.handleAuthChange(userId)
            }
        }
    }

    private func handleAuthChange(_ id: String?) {
        guard id != userId else { return }
        epoch += 1
        userId = id
        state = id.map { historyRepository.restore(userId: $0) } ?? .empty
    }

    func pay(billerId: String, billerName: String, categoryId: String, account: String, amountPaise: Int64) {
        if state.hasProcessing(billerId: billerId, account: account) { return }
        let payment = Payment(
            id: "pay-\(state.nextId)",
            billerId: billerId,
            billerName: billerName,
            categoryId: categoryId,
            account: account,
            amountPaise: amountPaise,
            paidAtUtc: clock.now(),
            status: .processing
        )
        state = state.adding(payment)
        events.emit(.paymentStarted(paymentId: payment.id))
        let startedEpoch = epoch
        pendingWork = Task {
            await persist()
            do {
                try await paymentRepository.pay(payment)
                if startedEpoch != epoch { return }
                state = state.updatingStatus(id: payment.id, status: .success)
                await persist()
                invalidateDueBills()
            } catch {
                if startedEpoch != epoch { return }
                state = state.updatingStatus(id: payment.id, status: .failed)
                await persist()
                events.emit(.paymentFailed(paymentId: payment.id))
            }
        }
    }

    private func persist() async {
        guard let id = userId else { return }
        do {
            try await historyRepository.persist(userId: id, value: state)
        } catch {
            events.emit(.storageFailed)
        }
    }
}
